templates/cc/string.go

```diff
@@ -4,64 +4,98 @@ const strTpl = `
-			return {{ err . "value does not have prefix " (lit $r.GetPrefix) }}
+	{
+		const std::string prefix = {{ lit $r.GetPrefix }};
+		if ({{ accessor . }}.compare(0, prefix.size(), prefix) != 0) {
+			{{ err . "value does not have prefix " (lit $r.GetPrefix) }}
 		}
+	}
 	{{ end }}
 
 	{{ if $r.Suffix }}
-		if !strings.HasSuffix({{ accessor . }}, {{ lit $r.GetSuffix }}) {
-			return {{ err . "value does not have suffix " (lit $r.GetSuffix) }}
+	{
+		const std::string suffix = {{ lit $r.GetSuffix }};
+		const std::string& value = {{ accessor . }};
+		if ((value.size() < suffix.size()) ||
+			(value.compare(value.size() - suffix.size(), suffix.size(), suffix) != 0)) {
+			{{ err . "value does not have suffix " (lit $r.GetSuffix) }}
 		}
+	}
 	{{ end }}
 
 	{{ if $r.Contains }}
-		if !strings.Contains({{ accessor . }}, {{ lit $r.GetContains }}) {
-			return {{ err . "value does not contain substring " (lit $r.GetContains) }}
+	{
+		if ({{ accessor . }}.find({{ lit $r.GetContains }}) == std::string::npos) {
+			{{ err . "value does not contain substring " (lit $r.GetContains) }}
 		}
+	}
 	{{ end }}
 
 	{{ if $r.GetIp }}
+		{{ unimplemented }}
+		{{/* TODO(akonradi) implement IP address constraints
 		if ip := net.ParseIP({{ accessor . }}); ip == nil {
 			return {{ err . "value must be a valid IP address" }}
 		}
+		*/}}
 	{{ else if $r.GetIpv4 }}
+		{{ unimplemented }}
+		{{/* TODO(akonradi) implement IP address constraints
 		if ip := net.ParseIP({{ accessor . }}); ip == nil || ip.To4() == nil {
 			return {{ err . "value must be a valid IPv4 address" }}
 		}
+		*/}}
 	{{ else if $r.GetIpv6 }}
+		{{ unimplemented }}
+		{{/* TODO(akonradi) implement IP address constraints
 		if ip := net.ParseIP({{ accessor . }}); ip == nil || ip.To4() != nil {
 			return {{ err . "value must be a valid IPv6 address" }}
 		}
+		*/}}
 	{{ else if $r.GetEmail }}
+		{{ unimplemented }}
+		{{/* TODO(akonradi) implement email address constraints
 		if err := m._validateEmail({{ accessor . }}); err != nil {
 			return {{ errCause . "err" "value must be a valid email address" }}
 		}
+		*/}}
 	{{ else if $r.GetHostname }}
+		{{ unimplemented }}
+		{{/* TODO(akonradi) implement hostname constraints
 		if err := m._validateHostname({{ accessor . }}); err != nil {
 			return {{ errCause . "err" "value must be a valid hostname" }}
 		}
+		*/}}
 	{{ else if $r.GetUri }}
+		{{ unimplemented }}
+		{{/* TODO(akonradi) implement URI constraints
 		if uri, err := url.Parse({{ accessor . }}); err != nil {
 			return {{ errCause . "err" "value must be a valid URI" }}
 		} else if !uri.IsAbs() {
 			return {{ err . "value must be absolute" }}
 		}
+		*/}}
 	{{ else if $r.GetUriRef }}
+		{{ unimplemented }}
+		{{/* TODO(akonradi) implement URI constraints
 		if _, err := url.Parse({{ accessor . }}); err != nil {
 			return {{ errCause . "err" "value must be a valid URI" }}
 		}
+		*/}}
 	{{ end }}
 
 	{{ if $r.MinLen }}
```

Comment:
So, this is the one we really need. I think we might either have to switch to using min bytes or figure out a way to get min length for Envoy data-plane-api. @mattklein123.

Reply:
I talked to @htuch in slack, and IMO all we actually need is "not_empty" or to be consistent with everything else, "required." I would just add that and not implement min_len. If that's not possible min_bytes is fine and we can set that to 1.